Add option for using compact properties for hashing

andremarianiello opened the pull request:
fixes https://github.com/centrifuge/precise-proofs/issues/29

lucasvo commented on proofs/property_name.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+type Property struct {
+	Parent  *Property
```
Could you use the same gofmt/golinting rules we use over here for this code: https://github.com/centrifuge/go-centrifuge/blob/develop/Makefile#L32

lucasvo replied:
I know this is still a WiP, adding it anyway just once as a reminder. The other thing that would be great is if you could add docstrings to all public interfaces.

andremarianiello replied:
Should I be basing this on the develop branch?

andremarianiello replied:
Oh sorry, that's another repo.

lucasvo replied:
Thanks! Feel free to copy the linter make command over into this repo :)

lucasvo commented on proofs/tree.go:
```diff
@@ -298,7 +299,7 @@ func (doctree *DocumentTree) GetLeafByProperty(prop string) (int, *LeafNode) {
 	propOrder := []string{}
 	for _, leaf := range doctree.leaves {
-		propOrder = append(propOrder, leaf.Property)
+		propOrder = append(propOrder, leaf.Property.Name())
```
This index should probably be duplicated for both types to allow lookup by human-readable values and bytes.

lucasvo replied:
Not quite. We use this index to create the proofs. I.e. given a property name, tell me which leaf in the tree a merkle proof should be created for. Given that we now support properties with human readable and compact version, we should have two indexes so that we can create a proof from either input. I think the right solution would be to create another method for the second index: 

`func (doctree *DocumentTree) CompactPropertyOrder() [][]byte`

lucasvo replied:
Actually []uint64 makes more sense.

andremarianiello replied:
@lucasvo Could I see an example of where `PropertyOrder` is used like you said? I changed this last night to return `[]Property` and I want to see if that fits your use-case, or if a separate function is really required. I think `[]Property` is better because you should be able to give either the compact or human-readable name, and then get the leaf node from that, and then compute the hash using either method, not necessarily the method that was used to find the node in the first place. But let me know if I'm wrong about that or if that use case doesn't make any sense.

lucasvo replied:
This method `DocumentTree.CreateProof` should support both formats. It uses it internally. Have you added a test case for creating a proof with the binary format already?

lucasvo commented on proofs/tree.go:
```diff
@@ -446,12 +460,11 @@ func (n *LeafNode) HashNode(h hash.Hash) error {
 	payload = append(payload, []byte(value)...)
 	if len(salt) != 32 {
-		return []byte{}, fmt.Errorf("%s: Salt has incorrect length: %d instead of 32", prop, len(salt))
+		return []byte{}, fmt.Errorf("%s: Salt has incorrect length: %d instead of 32", propBytes, len(salt))
```
This error isn't really human readable anymore if we use bytes but I guess there's no easy way to avoid it.

andremarianiello replied:
I didn't know whether it was better to leave it or remove it.

lucasvo replied:
Yeah, let's leave it unless go has a built in check to see if a type `[]byte` is a valid ASCII string. Then we could do an if/else but not worth adding any code just for this.

andremarianiello replied:
I have refactored ConcatValues to take a PropertyName. PropertyNames can be encoded as []byte or printed in a human-readable fashion. This leads to a nicely formatted error messages here in either case. Let me know what you think.

lucasvo replied:
Awesome!

lucasvo commented on proofs/tree.go:
```diff
@@ -753,21 +754,20 @@ func newMessageFlattener(message, messageSalts proto.Message, saltsLengthSuffix
+	var rootProp *Property
+	if parentPrefix != "" {
+		rootProp = &Property{
+			Text: parentPrefix,
```
This should be changed to allow supplying both a prefix for the "Text" and bytes.

lucasvo replied:
The challenge with the root property is that it only has an `OrdNum` of type `int64`. When creating a tree, we would want to add a prefix that is multiple levels deep. Eg.:

`document.data` => append(unt64(1), uint64(20))

It would be quite cumbersome to create a set of nested `Property` objects just to pass in these details. I think it's better to just ask for `[]uint64` instead.

andremarianiello replied:
If I change a Property to contain []uint64 instead of a single uint64, then you could specify an arbitrarily long prefix within a single property, and this would let this retain a Property as a prefix. Would you be okay with that? (Side benefit, this would get rid of the ugly *uint64 field in Property, which I think would be nicer as well).

andremarianiello replied:
I would still use nested properties to concat them, and keep the parent. I think the parent structure is useful.

lucasvo commented:
The PR looks good. There are three things that are still outstanding:
1) the utility to go from binary to human readable name and vice versa
2) test coverage for the binary format
3) a few comments above.

Could you avoid force pushing to the branch? It will make it a lot easier for me to review additional changes than having to go through all changes from scratch. Don't worry when merging, we'll squash & merge anyway. 

I've also added another comment on the issue worth checking out. #29  Not sure if you saw it.

andremarianiello commented on proofs/tree.go:
```diff
@@ -327,14 +326,9 @@ func (doctree *DocumentTree) CreateProof(prop string) (proof proofspb.Proof, err
+	propName := leaf.Property.Name(doctree.compactProperties)
 	proof = proofspb.Proof{
-		Property: propBytes,
+		Property: propName.AsBytes(),
```
Or it could be a LiteralPropName which is just a []byte. This would happen if we were calculating a hash from a Proof object. The Proof object has a []byte, not a PropertyName, so LiteralPropName is a simple way to create a PropertyName from the []byte (which gets converted back to a []byte on this line). We could store a PropertyName in the Proof object instead if you wanted to get rid of LiteralPropName.

lucasvo replied:
Not sure I'm following your explanation completely. The `Proof` struct is the export format (serialized to JSON to transmit in an API response for example).

It is used to by a third party concatenate the values of the leaf `(property, value, salt)` and validate the Merkle proof of the resulting hash against a known Merkle root.

 If the tree was constructed with the compact property names, the `Proof` needs to contain the compact property attribute. If it's using human readable properties, it should contain the string. As this value is needed when concatenating the leaf preimage.

I think the most elegant way would be to extend the `Property` protobuf message to have both `StringProperty` and `CompactProperty` or something similar and fill either value depending on the tree type.

andremarianiello replied:
> If the tree was constructed with the compact property names, the Proof needs to contain the compact property attribute. If it's using human readable properties, it should contain the string. As this value is needed when concatenating the leaf preimage.

Right now, the Proof contains only the bytes. In order to validate the Proof, one needs to know whether it is a compact or human-readable properties Proof. Everything currently works using that.

The question now is one that I can't answer because I don't have enough domain experience here, so I am fine doing whichever you prefer: should the Proof indicate whether its using compact properties or human-readable properties? Or should this be knowledge that is already known?

andremarianiello replied:
added the oneof property field to the proof protobuf

andremarianiello commented on proofs/tree.go:
```diff
@@ -289,22 +290,20 @@ func (doctree *DocumentTree) Generate() error {
 // GetLeafByProperty returns a leaf if it is found
 func (doctree *DocumentTree) GetLeafByProperty(prop string) (int, *LeafNode) {
```
Added

andremarianiello commented:
Added some basic PropertyName tests, and copied the basic CreateProof and modified it to test the compact hashes. Still looking into why that one isn't working.

andremarianiello commented on proofs/property.go:
```diff
@@ -0,0 +1,125 @@
+	return Property{
+		Parent:     &n,
+		Text:       "length",
+		NameFormat: SubFieldFormat,
```
Length properties intentionally have no length properties, unless you wish to encode them differently. I will fix the lack of test case though.

lucasvo replied:
Oh, if I understand you right:

```proto,
message Example {
    repeated int numbers = 1;
}
``` 

Would lead to the length being encoded in: `1` ? and the first element in `append(1, 0)`, second element in `append(1, 1)`?

That's fine with me, I don't see a drawback in this however it would be nice to document this behavior somewhere. Could you add a very short section to the [docstring](https://godoc.org/github.com/centrifuge/precise-proofs/proofs#hdr-Nested_and_Repeated_Structures)?

andremarianiello replied:
Yup, you understand correctly. I will update the docstring

andremarianiello replied:
Updated

andremarianiello commented on proofs/tree.go:
```diff
@@ -480,19 +493,7 @@ func (s LeafList) Swap(i, j int) {
-		}
-	}
-	return "", fmt.Errorf("Invalid protobuf annotation: %s", tag)
+	return strings.Compare(string(s[i].Property.ReadableName()), string(s[j].Property.ReadableName())) == -1
```
Sounds good. Should be easy to do.

lucasvo replied:
I'll put another bounty on it tomorrow. Don't have my ledger on me right now. Do you want to apply for #31 as well?

andremarianiello replied:
I will apply for that one if you prefer that I do it over someone else, but I don't mind if you want to approve someone else to work on it instead.

lucasvo commented:
> Is there a reason you prefer a simple bool of using a oneof field?
No, a oneof is fine.

lucasvo commented on proofs/tree.go:
```diff
@@ -332,16 +332,15 @@ func (doctree *DocumentTree) CreateProof(prop string) (proof proofspb.Proof, err
 	propName := leaf.Property.Name(doctree.compactProperties)
 	proof = proofspb.Proof{
-		Property: propName.AsBytes(),
+        Property: propName,
```
The spacing here is off.
